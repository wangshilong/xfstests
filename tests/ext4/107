#! /bin/bash
# FS QA Test No. 107
#
# Project quota.
# Use of "sync" mount option here is an attempt to get deterministic
# allocator behaviour.
#
#-----------------------------------------------------------------------
# Copyright (c) 2005 Silicon Graphics, Inc.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it would be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write the Free Software Foundation,
# Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
#-----------------------------------------------------------------------
#

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1	# failure is the default!
trap "rm -f $tmp.*; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/quota

_supported_fs ext4
_supported_os Linux #IRIX
_require_scratch

# real QA test starts here
_scratch_mkfs_ext4 -O quota,project >> $seqres.full 2>&1

# setup a default run
if [ -z "$MOUNT_OPTIONS" ]; then
	export MOUNT_OPTIONS="-o pquota,sync"
else
	export MOUNT_OPTIONS="$MOUNT_OPTIONS -o sync"
fi

_scratch_mount >> $seqres.full 2>&1

rm -f $tmp.projects
target=$SCRATCH_MNT/project
mkdir $target		|| exit
src/setproject -v 100 $target
chattr +P $target

echo "### deny a hard link - wrong project ID"
rm -f $SCRATCH_MNT/outer $target/inner
touch $SCRATCH_MNT/outer
src/setproject -v 99  $SCRATCH_MNT/outer
ln $SCRATCH_MNT/outer $target/inner 2>/dev/null
if [ $? -eq 0 ]; then
	echo hard link succeeded
	ls -ld $SCRATCH_MNT/outer $target/inner
else
	echo hard link failed
fi

echo "### allow a hard link - right project ID"
src/setproject -v 100  $SCRATCH_MNT/outer
ln $SCRATCH_MNT/outer $target/inner
if [ $? -eq 0 ]; then
	echo hard link succeeded
else
	echo hard link failed
	ls -ld $SCRATCH_MNT/outer $target/inner
fi

status=0
exit
