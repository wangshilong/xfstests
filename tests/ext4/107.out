QA output created by 107
### deny a hard link - wrong project ID
hard link failed
### allow a hard link - right project ID
hard link succeeded
